### PR TITLE
Add federation /user/keys/* endpoints

### PR DESCRIPTION
**Pull request by turt2live**
Rendered: see 'docs' status check

Honestly, I'm not too clear on how these APIs are used so there may be inaccuracies in the descriptions. 

----

Fixes https://github.com/matrix-org/matrix-doc/issues/1438

### CODE REVIEW

File: specification/server_server_api.rst [hunk 1]
```diff
@@ -890,6 +890,20 @@ nothing else.
 
 {{openid_ss_http_api}}
 
+End-to-End Encryption
+---------------------
+
+This section compliments the `End-to-End Encryption module`_ of the Client-Server
+API. For detailed information about end-to-end encryption, please see that module.
+
+The APIs defined here are designed to be able to proxy much of the client's request
```
Comment (ara4n) [inline specification/server_server_api.rst]: Hum. I think it's a bit more subtle than this, as the API needs to participate in the whole nightmare of synchronising devices between servers (as per https://github.com/matrix-org/matrix-doc/issues/1212). However, I'm also failing to spot how /users/devices ties into that, if at all. I suggest it would be worth putting a todo at this point though to say "MSC1212 goes here" though?

Comment (ara4n) [reply specification/server_server_api.rst]: (i have a bad feeling that /user/devices is used for blindly proxying the CS API over federation, as you say, but an entirely different API is used for determining a user's devices for E2E purposes :|)

Comment (turt2live) [reply specification/server_server_api.rst]: The client-server API doesn't appear to have a matching `/devices/{userId}` endpoint. I also just realized that devices are a whole thing that are unrelated to encryption, so I'll split that out. I have no idea what actually uses `/devices/{userId}` though.

**Comment by ara4n** [comment]
I'm torn between trying to reverse-engineer how this all works and hoping that @richvdh will have some bandwidth post-roomver stuff to unravel it.

File: specification/server_server_api.rst [hunk 2]
```diff
@@ -890,6 +890,31 @@ nothing else.
 
 {{openid_ss_http_api}}
 
+
+Device Management
+-----------------
+
+.. TODO: TravisR - Incorporate https://github.com/matrix-org/matrix-doc/issues/1212
```
Comment (turt2live) [inline specification/server_server_api.rst]: ftr - will do this in it's own PR once we figure out what this PR actually does.

File: specification/server_server_api.rst (hunk 1 above)
Comment (uhoreg) [reply specification/server_server_api.rst]: It looks like `/user/devices/{userId}` is called in response to a `m.device_list_update` EDU, if the server decides that it needs to resync.

Comment (turt2live) [reply specification/server_server_api.rst]: Which seems to make it part of #1212, if I understand it correctly. If that's the case, I'll split it out to it's own PR alongside the #1212 spec.

Comment (ara4n) [reply specification/server_server_api.rst]: uhoreg tells the truth. suggest we just move the /user/devices/{$userId} bit into the #1212 PR when it happens (which is what i'm looking into right now)

**Comment by turt2live** [comment]
I've taken out the stuff covered by #1212 - this should now be reviewable independently.

File: specification/server_server_api.rst [hunk 3]
```diff
@@ -1107,8 +1120,10 @@ that are too long.
 
 .. _`Invitation storage`: ../identity_service/unstable.html#invitation-storage
 .. _`Identity Service API`: ../identity_service/unstable.html
-.. _`Client-Server API`: ../client_server/unstable.html
+.. _`Client-Server API`: ../client_server/%CLIENT_RELEASE_LABEL%.html
 .. _`Inviting to a room`: #inviting-to-a-room
 .. _`Canonical JSON`: ../appendices.html#canonical-json
 .. _`Unpadded Base64`:  ../appendices.html#unpadded-base64
 .. _`Server ACLs`:  ../client_server/unstable.html#module-server-acls
+.. _`Device Management module`: ../client-server/%CLIENT_RELEASE_LABEL%.html#device-management
```
Comment (uhoreg) [inline specification/server_server_api.rst]: s/client-server/client_server/ (and in the next line as well)

File: api/server-server/user_keys.yaml [hunk 4]
```diff
@@ -0,0 +1,188 @@
...
+                  Information on the queried devices. A map from user ID, to a
+                  map from device ID to device information.  For each device,
+                  the information returned will be the same as uploaded via
+                  ``/keys/upload``, with the addition of an ``unsigned``
+                  property.
+                additionalProperties:
+                  type: object
+                  additionalProperties:
+                    allOf:
+                      - $ref: ../client-server/definitions/device_keys.yaml
```
Comment (uhoreg) [inline api/server-server/user_keys.yaml]: device_keys.yaml contains a link that needs to be defined at the bottom of server_server_api.rst.  The definition (from other files) is

```
.. _`Signing JSON`: ../appendices.html#signing-json
```

File: api/server-server/user_keys.yaml [hunk 5]
```diff
@@ -0,0 +1,188 @@
...
+        - signedRequest: []
+      parameters:
+        - in: body
+          name: body
+          type: object
+          required: true
+          schema:
+            type: object
+            properties:
+              device_keys:
```
Comment (uhoreg) [inline api/server-server/user_keys.yaml]: should be marked as required
